### ISSUE TITLE
Does flushing guarantee a reference count decrease?

### ISSUE THREAD

**Issue by nicmunroe** [netty/netty#6625]
Apologies if this is covered in documentation somewhere. I've looked around in Netty docs, javadocs, and done some google searching and couldn't find anything definitive. If documentation for this does exist then point me at it and I should be able to self-serve. If no docs exist covering this question then I guess this issue would represent my request to add docs covering it.

So here's the question: does flushing a channel guarantee that reference counted objects that were written to it will have their reference counts reduced by 1 eventually? i.e. 
```
// scenario 1 - write then flush
channel.write(someRefCountedMsg);
channel.flush();

// or scenario 2 - write and flush
channel.writeAndFlush(someRefCountedMsg);
```
I know that execution might be async depending on if you're in the correct event loop when those methods are called - I'm not worried about whether the ref count will decrease *immediately*, just if it's **guaranteed** to happen *eventually*.

In other words, is there a realistic scenario where the flush call fails and `someRefCountedMsg` is never `release()`d? I've looked in the code a bit and there does appear to be some codepaths where it might not be released if exceptions occur at certain points, but if that's never going to realistically happen even in the wild world of network communication where crazy stuff happens all the time, then I will happily not worry about it. 

But if there is a realistic scenario where flushing might not release reference counted objects then what is the recommended best practice for making sure I don't leak reference counted objects when flushing?

**Comment by ninja-** [comment]
Hi nicmunroe, 
the reference count is guaranted to be eventually decreased especially when either
1) the message is fully written to underlying socket (can take take some time under high load, it's not always happening on the first write(..) syscall )
2) the channel is suddeny closed, and the "buffered" not-yet-written messages are also released
3) exceptions happen, at least in most cases

What code paths did you find that could not release() messages? Are these some custom handlers?

**Comment by nicmunroe** [comment]
@ninja- Thanks for the response. That is encouraging - I want it to be the case that flush could never fail in a way that leaves dangling reference counts, at least not without a way to be notified of such a failure so I could release them myself, but given the way the code is structured and without a documented guarantee it made me want to double check here to be safe. For example this code in `AbstractChannelHandlerContext`:
```
    private void invokeFlush0() {
        try {
            ((ChannelOutboundHandler) handler()).flush(this);
        } catch (Throwable t) {
            notifyHandlerException(t);
        }
    }
```
There's try/catch logic to deal with thrown exceptions before the `release()` occurs (not just here, there are a few other places as well), and I don't know if it's there as a "this should never happen, but let's have some safety nets just-in-case" kind of thing, or a "this might happen but it will always happen after the reference count decrements", or if it's a realistic possibility that it might happen before ref count release in some scenarios. Thoughts?

(Side note: I haven't ever actually observed flush fail to decrement the message's reference count - this question is purely based on auditing my code for ways leaks *might* occur and bulletproofing against corner cases. If this is a possibility and there's a way to handle the failure then I want to know how to guard against it.)

**Comment by ninja-** [comment]
@nicmunroe at the code path you pasted there is nothing to fail. the message will be forwarded to the channel pipeline consisting of multiple handlers and it's each handler's responsibility to always call .release(), even in case of exception.
You can see an example here: https://github.com/netty/netty/blob/4.1/codec/src/main/java/io/netty/handler/codec/MessageToMessageDecoder.java#L89

**Comment by nicmunroe** [comment]
@ninja- that's good general advice for inbound messages, and I understand it. But in this case since it's outbound, I'm not supposed to be releasing in my code. From what I understand, on outbound stuff it's the job of the Netty handler to release it as stated [here](http://netty.io/wiki/reference-counted-objects.html#user-content-outbound-messages) in the "Outbound messages" section of the reference-counting wiki page.

I'm not worried about my code throwing exceptions - I've put finally blocks in and otherwise covered my code correctly. But in the case of flushing an outbound message via the built-in Netty `Channel.flush()` or `Channel.writeAndFlush()` I'm concerned about if something can go wrong before the message is `release()`d given how many things can go wrong when dealing with network communication. 

I'm leaning towards simply trusting that flush will *always* release a message no matter what given what it says in that wiki section and what you mentioned earlier, but I was hoping someone with deep knowledge of that area of Netty code could confirm that there's no real-life scenario that would cause `Channel.flush()` to fail to `release()` a reference counted object that had been written to the channel.

**Comment by ninja-** [comment]
flush() is not doing much by itself, majority of the job is done by write() which is passing the message via each of the encoders.

so as someone with deep knowledge of that area of Netty code I can confirm that there's no real-life scenario that would cause Channel.flush() to fail to release() a refrence counted object that had been written to the channel...:P

**Comment by nicmunroe** [comment]
Perfect, thank you! Sorry for hammering on it - I don't think I did a good job of explaining myself. And thanks for the quick replies!

**Comment by ninja-** [comment]
Always happy to help 😄
